0000-ldap-sync-tool.md

```diff
@@ -0,0 +1,233 @@
+STDOUT, if requested.
+
+# Uyuni UI Enhancements
+
+Considering that this tool a new way of creating/modifying users
+beside the UI and API in an automated way, there should be an API
+endpoint that allows to set a flag in User Management that gives
+admins the clear indication that the certain user is managed externally.
+
+Main reason is the role mapping that is included. An admin currently
+easily can change the user settings in the UI and will be wondering
+why his changes get reset to the original settings. Therefore, Uyuni
+UI should mark external users either with special icon or color etc
+and the page of such user should have disabled "Update" button and an
+explanation for such user is managed elsewhere externally.
+
+# Reference Implementation
+
+This is already implemented in the prototype which is available at 
+[uyuni-ldap-sync repository](https://github.com/isbm/uyuni-ldap-sync).
+
+Current prototype has two releases:
+1. [v0.1](https://github.com/isbm/uyuni-ldap-sync/releases/tag/v0.1) just reimplements
+[existing script in Uyuni](https://github.com/uyuni-project/uyuni/blob/master/utils/sw-ldap-user-sync).
+2. [v0.2](https://github.com/isbm/uyuni-ldap-sync/releases/tag/v0.2) implements the current RFC and adds Active Directory support
+
+Moving from prototype status to a production release, the following
+would be still required:
+
+- Add unit tests
+- Possible refactorings
+
+## Additional Features
+
+Features that might come with the next release:
+
+- Automatic (transparent) PAM authentication setup/un-setup
+- Daemon mode
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+N/A
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Native Implementation
+
+Alternative design would be to extend user management in the very
+Uyuni server, making it LDAP-aware natively. However, the amount of
+changes to the existing code might be very large.
+
+## Salt-based Implementation
+
+It could be possible to reuse existing Spacewalk runner in Salt,
+wrapping it into a state module and implement all that as a Salt
+Formula. While such approach would give more flexibility to the user,
+it will also have a number of drawbacks, such as:
+
+- Additional formula debugging difficulties overhead
+
+- User-introduced bugs that might sync LDAP wrong with severe
+  consequences 
+
+- Fragility of the very solution, since it is still an editable
+  Formula
+
+- No daemon mode. Every time LDAP database update, changes supposed to
+  be just propagated transparently. For that, the job in Salt should
+  be scheduled, and thus calling Salt locally would be too much of overhead
+
+Salt, however, could periodically call `uyuni-ldap-sync` and/or setup
+configuration for it (PAM, the `uyuni-ldap-sync` configuration etc).
+
+With daemon mode implemented, Salt supposed to only setup the
+`uyuni-ldap-sync` once and delegate the rest to the
+`uyuni-ldap-sync`.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+N/A
```

Comment:
Context: SSO via SAML.
Can we use the `uyuni-ldap-sync` tool create an on-the-fly user that is authenticated in the SSO but not existing in the Uyuni user database and let him/her login?

Reply:
Uhm, but how would you imagine that? The SAML part is a completely different approach.

The current sync does exactly how it is designed for and how it is called: _synchronising_ LDAP data. Same applies to moio's Salt-based idea, except it is just slower and much more fragile. Basically, syncing works by just duplicating everything into local DB of the Uyuni from the LDAP, so the *standard* Uyuni's *built-in* auth routines will apply on request and issue a session.

Now, if you want single sign on via SAML (or oAuth), you still want to have an identity provider somewhere that will sign the responses for you, which you will process directly in Tomcat. That said, the user is authenticated elsewhere (at the provider) and provider sends your app (the Uyuni on Tomcat) a signed message for "yes, let him in" or "no, kick him away", as there is no longer any user database available at the Uyuni side per se (hence why it is called SSO, not redistributed DB of users).

Hope that clarifies.

Reply:
> Now, if you want single sign on via SAML (or oAuth), you still want to have an identity provider somewhere that will sign the responses for you, which you will process directly in Tomcat. That said, the user is authenticated elsewhere (at the provider) and provider sends your app (the Uyuni on Tomcat) a signed message for "yes, let him in" or "no, kick him away", as there is no longer any user database available at the Uyuni side per se (hence why it is called SSO, not redistributed DB of users).

Yes, this is what I was referring to: the SAML gets back with "authentication for $username successful". 
The question is: do you think we can reuse part of the solution you devised to create an on-the-fly user in the Uyuni/SUSE Manager database (that we mandatory need - think of user roles) upon successful login message from the SAML entity (IdP)?

Reply:
As per conversation in chat, seems I was misled by the original question formulation, so we lost common ground:

> Can we use the uyuni-ldap-sync tool create an on-the-fly user that is authenticated in the SSO **but not existing in the Uyuni user database** and let him/her login?

The correct question would be:

> Can we use `uyuni-ldap-sync` tool to pre-sync users in the local database so they can be authorised after being authenticated with SAML?

In that case, yes.